crates/flux-refineck/src/checker.rs

```diff
@@ -229,20 +231,70 @@ impl<'ck, 'genv, 'tcx> Checker<'ck, 'genv, 'tcx, RefineMode> {
 /// `SubFn` lets us reuse _most_ of the same code for `check_fn_subtyping` for both the case where
 /// we have an early-bound function signature (e.g., for a trait method???) and versions without,
 /// e.g. a plain closure against its FnTraitPredicate obligation.
+#[derive(Debug)]
 pub enum SubFn {
-    Poly(EarlyBinder<rty::PolyFnSig>, rty::GenericArgs),
+    Poly(DefId, EarlyBinder<rty::PolyFnSig>, rty::GenericArgs),
     Mono(rty::PolyFnSig),
 }
 
 impl SubFn {
     pub fn as_ref(&self) -> &rty::PolyFnSig {
         match self {
-            SubFn::Poly(sig, _) => sig.skip_binder_ref(),
+            SubFn::Poly(_, sig, _) => sig.skip_binder_ref(),
             SubFn::Mono(sig) => sig,
         }
     }
 }
 
+/// The function `hoist_with_binders` uses the `LocalHoister` machinery to convert
+/// a (closure) function template _without_ binders, e.g. `fn ({v.i32 | *}) -> {v.i32|*})`
+/// into one _with_ input binders, e.g. `forall <a:int>. fn ({i32[a]|*}) -> {v.i32|*}`
+/// after which the hole-filling machinery can be used to fill in the holes.
+/// This lets us get "dependent signatures" where the output can refer to the input.
+/// e.g. see `tests/pos/surface/closure09.rs`
+fn hoist_input_binders(poly_sig: &PolyFnSig) -> PolyFnSig {
```

Comment:
Cole needed the same so he implemented it here https://github.com/flux-rs/flux/blob/12d42d997ac645495eb0e44774f3dddf96116519/crates/flux-middle/src/rty/canonicalize.rs#L229. We should reuse that. You'll need to modify it to collapse holes.

Reply:
done!